QEMU Workflow for Throw/Catch Validation

BigBoySanchez opened the pull request:
## Overview

This PR adds a basic QEMU simulation workflow for running and testing the exception runtime. The goal is to verify correct memory behavior (e.g., variable writes) after a `throw`/`catch`

## What it does

- Verify that the "global" variable is set to the right value for both demos (single_level & multiple_inheritance)
- Verify both demos work for stm32f103c8 and lpc4078
- Avoid clock configuration when building for debug

## Changes

### ci.yml
- Calls demo_check, instead of build_demo for jobs demo_check_stm32f103c8 and demo_check_lpc4078

### demo_check.yml
- olimex-stm32-h405 simulated stm32f103c8 and lpc4078 well, so I used that as the machine on qemu-system-arm
- Builds the demos in release and debug
- For both demos, run qemu with gdb to verify that the variable "global" is set to the right value

### demos/conanfile.py
- Added code to create a macro that disables clock configuration for QEMU tests
- Macro is created when build_type = debug
- Dependency code is there to solve bugs that appeared after implementing macro logic

### demos/platforms/lpc4078.cpp
- Wrapped `hal::lpc40:maximum` in an `#ifndef DEBUG` guard so that it’s excluded for QEMU testing

### demos/platforms/stm32f103c8.cpp
- Wrapped `hal::stm32f1::maximum_speed_using_internal_oscillator` in an `#ifndef DEBUG` guard so that it’s excluded for QEMU testing

---

🧪 **You can view sample workflow runs under `Actions → ci`** in this repo.

BigBoySanchez committed:
testing qemu in actions

BigBoySanchez committed:
qemu on actions doesn't have right machine

BigBoySanchez committed:
using ubuntu latest for qemu

BigBoySanchez committed:
oops forgot to update apt

BigBoySanchez committed:
oops forgot to update apt

BigBoySanchez committed:
trying to run elf in qemu

BigBoySanchez committed:
using qemu + gdb to verify memory

BigBoySanchez committed:
added debug logic

BigBoySanchez committed:
fixed artifact issue, added logs to gdb

BigBoySanchez committed:
cleaning up .yml file

BigBoySanchez committed:
Merge pull request #1 from libhal/main

Update Main

kammce commented on .github/workflows/demo_check.yml:
```diff
@@ -0,0 +1,186 @@
+      - name: Check Single Level Result
+        run: |
+          cat gdb.log
+        
```
```suggestion
```

Lets remove this extra line

kammce commented on .github/workflows/demo_check.yml:
```diff
@@ -0,0 +1,186 @@
+            -ex "set output-radix 16"                                                                           \
+            -ex "target remote :1234"                                                                           \
+            -ex "watch global"                                                                                  \
+            -ex "continue"                                                                                      \
+            -ex "monitor quit"                                                                                  \
```
What happens if we never update global? Is there a way to set a timeout with continue? Otherwise our CI may get stuck and we end up wasting a lot of server time. Same goes for  the other test.

kammce replied:
We could add the `timeout` command before the `gdb-multiarch` and that should do that for us. I'm thinking like 5 seconds is enough.

kammce reviewed (changes requested):
Great job @BigBoySanchez ! This is a great contributiont to libhal-exceptions.

BigBoySanchez committed:
removed empty line between cat and grep

BigBoySanchez committed:
testing bg process interactions in actions

BigBoySanchez committed:
test an unchanged global scenario

BigBoySanchez committed:
added logs with explicit output

BigBoySanchez commented:
## Edits
- Got rid of the empty line between `cat` and `grep` for both checks
- Added `timeout` of 5s to `gdb-multiarch` for both checks
- Timeout does have a distinct exit code, but I added extra code to make timeouts easier to detect from the action logs

## Sample runs
- Timeout by running w/o a watchpoint: [#42](https://github.com/BigBoySanchez/libhal-exceptions/actions/runs/16617903600)
- Fully functional run: [#44](https://github.com/BigBoySanchez/libhal-exceptions/actions/runs/16618373458)
---
Thank you so much for taking the time to review my code 😊

kammce reviewed (approved):
Looks great. Make a PR on pi half exceptions but sqaush the commits down to just 1 rebased off main. Great work 😁
